src/ripple/app/misc/detail/WorkBase.h

```diff
@@ -182,8 +195,30 @@ WorkBase<Impl>::onResolve(error_code const& ec, resolver_type::iterator it)
     if (ec)
         return fail(ec);
 
+    resolver_type::iterator end;
+    std::vector<
+        std::reference_wrapper<resolver_type::iterator::value_type const>>
+        entries;
```

Comment:
I can see why this is written this way. It's a straight forward way to choose an random endpoint that was not previously failing. However, this code creates a vector or entries to randomly choose from, even when we won't be choosing from the vector. This will allocate memory, which we try to avoid. Even when we will be making a random choice, I don't think we need to create a vector to do so. What do you think about this refactoring?
```
template <class Impl>
void
WorkBase<Impl>::onResolve(error_code const& ec, resolver_type::iterator it)
{
    if (ec)
        return fail(ec);

    lastEndpoint_ = [&]() -> endpoint_type {
        // Use last endpoint if it is successfully connected and is in the list,
        // otherwise pick a random endpoint from the list. If there is only one
        // endpoint and it is the last endpoint then use the last endpoint.

        resolver_type::iterator end;
        auto const foundIt = std::find(it, end, lastEndpoint_);
        if (lastStatus_ && foundIt != end)
            return lastEndpoint_;

        switch (auto const n = std::distance(it, end); n)
        {
            case 0:
                return lastEndpoint_;
            case 1:
                return *it;
            default: {
                // need to choose a random element from the collection that does
                // not equal `lastEndpoint_`` if `lastEndpoint_` is part of the
                // collection.
                if (foundIt == end)
                    return *std::next(it, rand_int(n - 1));
                // lastEndpoint_ is part of the collection
                // Pick a random number from the n-1 valid choices, if we use
                // this as an index, note the last element will never be chosen
                // and the `lastEndpoint_` index may be chosen. So when the
                // `lastEndpoint_` index is chosen, that is treated as if the
                // last element was chosen.
                auto randIndex = (n > 2) ? rand_int(n - 2) : 0;
                auto const foundIndex = std::distance(it, foundIt);
                if (randIndex == foundIndex)
                    randIndex = n - 1;
                return *std::next(it, randIndex);
            }
        }
    }();

    socket_.async_connect(
        lastEndpoint_,
        strand_.wrap(std::bind(
            &Impl::onConnect,
            impl().shared_from_this(),
            std::placeholders::_1)));
}
```

Reply:
FWIW, when I saw the `vector` being populated and usually not used I had concerns similar to those expressed by @seelabs.  I'd prefer an approach which either does not use the `vector` or only populates the `vector` if it will actually use the contents.